File: Exercise_1.py
# Problem1 (https://leetcode.com/problems/binary-search-tree-iterator/)
# // Time Complexity :O(1)
# // Space Complexity :O(1)
# // Did this code successfully run on Leetcode :Yes
# // Any problem you faced while coding this :No


# // Your code here along with comments explaining your approach

class BSTIterator:
    def __init__(self, root):
        # creating a stack to store all elements of the BST
        self.stack = []
        # applying dfs
        self.dfs(root)

    def dfs(self, root):
        # iterate to the leftmost leaf node
        while root:
            # keep appending to stack
            self.stack.append(root)
            root = root.left

    def next(self):
        # pop the last element from the stack
        curr = self.stack.pop()
        # when popped iterate to the right most leaf node
        self.dfs(curr.right)
        return curr.val

    def hasNext(self):
        return self.stack != []

File: Exercise_1.cpp
// Time Complexity                              :  next() - O(1) , hasNext() - O(1), initialization takes O(N) time.
// Space Complexity                             :  O(N) - number of nodes. The values of the nodes are stored in a queue.
// Did this code successfully run on Leetcode   :  Yes
// Any problem you faced while coding this      :  No

#include <bits/stdc++.h> 
using namespace std;  

// https://leetcode.com/problems/binary-search-tree-iterator/submissions/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class BSTIterator {
private: 
    queue<int> q;
    void inorder(TreeNode *root) {
        if(!root) return;
        inorder(root->left);
        q.push(root->val);
        inorder(root->right);
    }
public:
    
    BSTIterator(TreeNode* root) {
        inorder(root);
    }
    
    int next() {
        int nval = q.front();
        q.pop();
        return nval;
        
    }
    
    bool hasNext() {
        return !q.empty();
    }
};

/**
 * Your BSTIterator object will be instantiated and called as such:
 * BSTIterator* obj = new BSTIterator(root);
 * int param_1 = obj->next();
 * bool param_2 = obj->hasNext();
 */